src/metrics.jl

```diff
@@ -239,7 +239,7 @@ corr_dist(a::AbstractArray, b::AbstractArray) = evaluate(CorrDist(), a, b)
 result_type(::CorrDist, a::AbstractArray, b::AbstractArray) = result_type(CosineDist(), a, b)
 
 # ChiSqDist
-@inline eval_op(::ChiSqDist, ai, bi) = abs2(ai - bi) / (ai + bi)
+@inline eval_op(::ChiSqDist, ai, bi) = ifelse(ai != bi, abs2(ai - bi) / (ai + bi), zero(ai))
```

Comment:
Shouldn't the check be `ifelse(ai + bi != 0, ...)`? Also, it's not easy to ensure type stability (e.g. for integer inputs), maybe use something like `zero(ai - bi)/one(ai - bi)`?

Reply:
> Shouldn't the check be `ifelse(ai + bi != 0, ...)`?

We can define it as 0 only at `(0, 0)`.
For the neighbourhood of `(x, -x)` line (points like (x, -x + z), where z -> 0) the function is `(2x+z)^2 / z`, i.e. either -> -Inf or -> +Inf depending on `z` sign. So I think it should stay NaN everywhere except (0,0) where it behaves as `z`.

> maybe use something like `zero(ai - bi)/one(ai - bi)`?

+1